upgrade/helpers/tasks.py

```diff
@@ -1515,3 +1516,134 @@ def add_satellite_subscriptions_in_capsule_ak(ak):
                                 f'"{output}" --id="{ak.id}"')
                 if ak_output.return_code > 0:
                     logger.warn(output)
+
+
+def pulp2_pulp3_migration():
+    """
+    Use to perform the pulp migration before running the upgrade.
+    """
+    def estimation():
+        """
+        Use to calculate the approximate time for migration.
+        :return: migration-stats status code
+        """
+        logger.info(
+            "checking the approximate migration time before starting the pulp migration")
+        with fabric_settings(warn_only=True):
+            output = run("satellite-maintain content migration-stats")
+            logger.info(output)
+            if output.return_code != 0:
+                logger.warn(output)
+        return output.return_code
+
+    def preparation():
+        """
+        Use to prepare the pulp migration environment.
+        :return: prep-<to_version>-upgrade status code
+        """
+        logger.info("starting the pulp pre-migration to change the ownership, permissions "
+                    "of the pulp content")
+        with fabric_settings(warn_only=True):
+            preup_time = datetime.now().replace(microsecond=0)
+            output = run(f"satellite-maintain prep-{settings.upgrade.to_version}-upgrade")
+            postup_time = datetime.now().replace(microsecond=0)
+            logger.info(output)
+            logger.highlight(f"pulp2-pulp3 pre-migration completed successfully and it took: "
+                             f"{str(postup_time - preup_time)}")
+            if output.return_code != 0:
+                logger.warn(output)
+            return output.return_code
+
+    def migration():
+        """
+        Use to run the pulp2 to pulp3 migration step.
+        :return: satellite-maintain content prepare
+        """
+        if bz_bug_is_open(1967131):
+            output = run('sed -i "s/downloaded = record.downloaded if '
```

Comment:
This piece of code is used to apply the patch (BZ#1967131) so that we can avoid pulp migration failure, and it will be applied until BZ gets not fixed.